### PR TITLE
Two step migration

### PR DESCRIPTION
**Pull request by michamos**
## Description

Changes migrator to use two-step migration, i.e. populating the mirror first (`inspire_prod_records` table) and migrating afterwards. This also introduces a new `migrate` cli which makes more sense.

## Checklist:

- [ ] I have all the information that I need (if not, move to `RFC` and look for it).
- [ ] I linked the related issue(s) in the corresponding commit logs.
- [ ] I wrote [good commit log messages](https://github.com/torvalds/subsurface-for-dirk/blob/5f15ad5a86ada3c5e574041a5f9d85235322dabb/README#L92-L119).
- [ ] My code follows the code style of this project.
- [ ] I've added any new docs if API/utils methods were added.
- [ ] I have updated the existing documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by michamos** [comment]
Still misses some tests for the new CLI.

File: inspirehep/modules/migrator/cli.py [hunk 1]
```diff
@@ -82,31 +150,30 @@ def populate(file_input=None,
...
         click.echo("Migrating records from file: {0}".format(file_input))
 
-        migrate(os.path.abspath(file_input), wait_for_results=wait)
+        migrate_from_file(os.path.abspath(file_input), wait_for_results=wait)
 
 
 @migrator.command()
 @click.option('--recid', '-r', type=int, help="recid on INSPIRE")
 def one(recid):
+    """DEPRECATED Migrate one record from legacy."""
```
Comment (puntonim) [inline inspirehep/modules/migrator/cli.py]: Maybe you could use a proper warning: https://docs.python.org/2/library/warnings.html

File: inspirehep/modules/migrator/ext.py [hunk 2]
```diff
@@ -33,5 +33,6 @@ def __init__(self, app=None):
             self.init_app(app)
 
     def init_app(self, app):
+        app.cli.add_command(migrate)
         app.cli.add_command(migrator)
```
Comment (puntonim) [inline inspirehep/modules/migrator/ext.py]: Add a comment here saying that migrator is deprecated.

File: inspirehep/modules/migrator/models.py [hunk 3]
```diff
@@ -52,3 +56,29 @@ def marcxml(self):
...
     def marcxml(self, value):
         self._marcxml = compress(value)
+
+    @hybrid_property
+    def error(self):
+        return self._errors
+
+    @error.setter
+    def error(self, value):
+        """errors column setter that stores an Exception and sets the ``valid`` flag."""
```
Comment (drjova) [inline inspirehep/modules/migrator/models.py]: ```s/errors/Errors/```

File: tests/integration/conftest.py [hunk 4]
```diff
@@ -27,6 +27,12 @@
 
 import pytest
 
+from functools import partial
+
+from click.testing import CliRunner
+from flask import current_app
+from flask_cli import ScriptInfo
```
Comment (drjova) [inline tests/integration/conftest.py]: ```from flask.cli import ScriptInfo```

``flask_cli`` has been deprecated https://pythonhosted.org/Flask-CLI/

File: inspirehep/modules/migrator/models.py [hunk 5]
```diff
@@ -52,3 +56,29 @@ def marcxml(self):
...
+    @classmethod
+    def from_marcxml(cls, raw_record):
+        """Create an instance from a MARCXML record.
+
+        The record must have a ``001`` tag containing the recid, otherwise it raises a ValueError.
+        """
+        try:
+            recid = int(cls.re_recid.search(raw_record).group('recid'))
+        except (ValueError, AttributeError):
+            raise ValueError('The MARCXML record contains no recid or recid is malformed')
```
Comment (drjova) [inline inspirehep/modules/migrator/models.py]: What about a more specific error such as ``MalfromedMARCXML`` or similar?  WDYT?

File: inspirehep/modules/migrator/cli.py [hunk 6]
```diff
@@ -64,9 +65,76 @@
 )
 
 
+def halt_if_debug_mode(force):
+    if not force and current_app.config.get('DEBUG'):
+        click.echo('The application is running in debug mode, which leaks memory when doing many database operations. To avoid problems, disable debug mode. This can be done by setting "DEBUG=False" in the config or setting the environment variable "APP_DEBUG=False". If you know what you are doing, you can pass the "--force" flag to disable this check.', err=True)
```
Comment (drjova) [inline inspirehep/modules/migrator/cli.py]: this line is too looooong 😄

File: inspirehep/modules/migrator/tasks.py [hunk 7]
```diff
@@ -366,43 +414,46 @@ def record_insert_or_replace(json, skip_files=False):
...
-        _store_migrator_error(recid, raw_record, e)
+        prod_record.error = exc
+        db.session.merge(prod_record)
         return None
 
     if '$schema' in json_record:
         ensure_valid_schema(json_record)
 
     try:
         record = record_insert_or_replace(json_record, skip_files=skip_files)
```
Comment (drjova) [inline inspirehep/modules/migrator/tasks.py]: This should be enough to be compatible with the centralized record creation :)
```python
from inspirehep.modules.records.api import InspireRecord

record = InspireRecord.create_or_update(json_record, skip_files=skip_files)
record.commit()
```

File: inspirehep/modules/migrator/models.py (hunk 5 above)
Comment (michamos) [reply inspirehep/modules/migrator/models.py]: I don't think it's very useful, as this is not something that should ever happen, it's more of a sanity check.

File: inspirehep/modules/migrator/tasks.py [hunk 8]
```diff
@@ -332,15 +369,41 @@ def _get_records_to_update_generator(citations_lookup):
...
 
 
+def insert_into_mirror(raw_records):
+    try:
+        for raw_record in raw_records:
+            prod_record = InspireProdRecords.from_marcxml(raw_record)
+            db.session.merge(prod_record)
+        db.session.commit()
+    finally:
+        db.session.close()
```
[the remark this answers is not in the document]
Comment (puntonim) [reply inspirehep/modules/migrator/tasks.py]: Also: https://github.com/michamos/inspire-next/blob/35199b5a1c69b0917690244aa5e0c74ec3792446/inspirehep/modules/migrator/tasks.py#L289

**Pull request merged**